tests/pass/move-data-across-await-point.rs

```diff
@@ -0,0 +1,84 @@
+use std::future::Future;
+
+// This test:
+// - Compares addresses of non-Copy data before and after moving it
+// - Writes to the pointer after it has moved across the await point
+//
+// This is only meant to assert current behavior, not guarantee that this is
+// how it should work in the future. In fact, upcoming changes to rustc
+// *should* break these tests.
+// See: https://github.com/rust-lang/rust/issues/62958
+async fn data_moved_async() {
+    async fn helper(mut data: Vec<u8>, raw_pointer: *mut Vec<u8>) {
+        let raw_pointer2 = &mut data as *mut Vec<u8>;
+        // When the Vec is moved into the generator upvar, that copies the Vec's
+        // ptr+capacity+len that is on the stack. Because of that, the value of
+        // `raw_pointer2` is different than that of `raw_pointer`.
+        //
+        // When copy propagation is enabled for generator upvars, the pointer
+        // addresses should be equal, becouse there will no longer be a
+        // generator upvar; instead, the function's `data` argument will be
+        // used directly.
+        assert_ne!(raw_pointer, raw_pointer2);
+        unsafe {
+            // This writes into the `x` in `data_moved_async`, re-initializing it.
+            std::ptr::write(raw_pointer, vec![3]);
+        }
+    }
+    // Vec<T> is not Copy
+    let mut x: Vec<u8> = vec![2];
+    let raw_pointer = &mut x as *mut Vec<u8>;
```

Comment:
```suggestion
    let raw_pointer = ptr::addr_of_mut!(x);
```
Also needs a `use std::ptr;`. Should be done for all the `&mut _ as *mut _` in this file.